fix(qna) : implemented basic features to the new codebase

ierezell opened the pull request:
Added filter, enable disable, get number of qna, export import. 

Import import still work per Qna but the logic is here and can be copied in the export topic as a sub logic.

MaxCloutier committed:
feat(nlu): add intent form to nodes

MaxCloutier committed:
Add placeholder translations and other fixes

MaxCloutier committed:
Add expand button

MaxCloutier committed:
Fix icons and styles

MaxCloutier committed:
Merge branch 'dev' into max-intent-form

MaxCloutier committed:
Fix PR

MaxCloutier committed:
Fix data

MaxCloutier committed:
Merge branch 'dev' into max-intent-form

MaxCloutier committed:
Fix dependency issue

MaxCloutier committed:
Fix super input array issues

MaxCloutier committed:
Put git hash back

MaxCloutier committed:
Fix stuff

MaxCloutier committed:
Merge branch 'dev' into max-intent-form

MaxCloutier committed:
Remove post-processing of tags as it's been fixed in another PR

MaxCloutier committed:
Merge branch 'max-intent-form' of github.com:botpress/botpress into max-intent-form

slvnperron committed:
Merge branch 'dev' of github.com:botpress/botpress into max-intent-form

slvnperron committed:
train from flows intents working

slvnperron committed:
implemented complex types

allardy committed:
merge dev

allardy committed:
yarn lock

slvnperron committed:
fixed PR comments

slvnperron committed:
Merge branch 'max-intent-form' of github.com:botpress/botpress into sp_new-intents

slvnperron committed:
partial progress towards qna refac

slvnperron committed:
qna responses works

slvnperron committed:
fix(nlu): remove empty tokens

slvnperron committed:
refactoring

slvnperron committed:
built-in intents moved to .intents.json

slvnperron committed:
builtin email pattern

Commit:
fix(ndu) :qna => filter and enable fixed

Commit:
number of question by topics done

Commit:
Delete launch.json

Commit:
removed one consolelog

Commit:
Merge branch 'ps_fix_sp_refac_qna' of github.com:botpress/botpress into ps_fix_sp_refac_qna

Commit:
readed launch.json

Commit:
fix to lwercase for search

Commit:
partial progress for import export

Commit:
fix(qna) : fixing qnas from refactor (#3802)

* fix(ndu) :qna => filter and enable fixed

* number of question by topics done

* Delete launch.json

* removed one consolelog

* readed launch.json

* fix to lwercase for search

* partial progress for import export

Commit:
Merge branch 'sp_refac-qna' of github.com:botpress/botpress into ps_fix_sp_refac_qna

Commit:
started to work on zip to export import

Commit:
import export functional

Commit:
cleaned code

Ierezell commented on modules/qna/src/backend/api.ts:
```diff
@@ -4,9 +4,19 @@ import moment from 'moment'
+  topicName: string
+  botId: string,
+  zipFile: Buffer
+  bpCms: typeof sdk.cms
```
removed bpCms  as it was not used at the end

Ierezell commented on modules/qna/src/backend/storage.ts:
```diff
@@ -1,10 +1,16 @@
+import mkdirp from 'mkdirp'
 import nanoid from 'nanoid/generate'
 import path from 'path'
+import { BPFS_STORAGE } from 'process'
```
removed (I guess it was old code)

Ierezell commented on modules/qna/src/views/full/Components/ImportModal.tsx:
```diff
@@ -159,58 +128,6 @@ export const ImportModal: FC<Props> = props => {
-      </Fragment>
-    )
-  }
-
```
This was a point to discuss, The base logic is made but changes will be made regarding the UX

EFF replied:
I think we can at least warn the user that the import will override conflicting qnas but what we don't want is to show a 2nd step like the one we had in the modal

Ierezell commented on modules/qna/src/views/full/index.tsx:
```diff
@@ -159,21 +159,21 @@ const QnAList: FC<Props> = ({
+      tooltip: lang.tr('importJson')
+    }
+  )
+  // }
```
removed but cannot test without them, condition is always false so export/import button is never shown.

Commit:
requested changes

EFF commented on modules/qna/src/backend/api.ts:
```diff
@@ -102,32 +110,35 @@ export default async (bp: typeof sdk, bots: ScopedBots) => {
     try {
-      const importData: Item[] = JSON.parse(req.file.buffer)
-      await Promise.map(importData, q => storage.updateSingleItem(req.query.topicName, q))
+      const importArgs: ImportArgs = {
```
we still need to decide on what logic we want to impose to users here

EFF commented on modules/qna/src/backend/storage.ts:
```diff
@@ -139,29 +150,103 @@ export default class Storage {
+      return { ...i, metadata: newMetadata }
+    })
+    const contentToExport: sdk.FormData[] = _.flatMapDeep(jsonQna.map(i => i.metadata.contentAnswers))
+    const mediaToExport = _.flatMapDeep(contentToExport.map(c => c.items ? c.items.map(i => i.image) : c.image))
```
you can replace this with a single chainable

```ts
const medias = _.chain(jsonQna)
 .map(item => item.metadata?.contentAnswers ?? [])
 .flatMap(c => c.items ? c.items : [c])
 .map(c => c.image)
 .value()
```

EFF commented on modules/qna/src/backend/storage.ts:
```diff
@@ -139,29 +150,103 @@ export default class Storage {
+    await mkdirp.sync(path.join(tmpDir.name, MEDIA_FOLDER))
+
+    const jsonQnaForBot = await this.ghost.readFileAsObject<Intent[]>(FLOW_FOLDER, toQnaFile(topicName))
+    const jsonQna = jsonQnaForBot.map(i => {
+      const newContent = [...i.metadata.contentAnswers].map(c => removeBotPrefix(c))
+      const newMetadata = { ...i.metadata, contentAnswers: newContent }
+      return { ...i, metadata: newMetadata }
+    })
```
```suggestion
    const jsonQna = jsonQnaForBot.map(i => {
      const contentAnswers = [...i.metadata?.contentAnswers ?? []].map(removeBotPrefix)
      const metadata = { ...i.metadata, contentAnswers }
      return { ...i, metadata }
    })
```

Ierezell replied:
I was me who changed some Sync functions to async and I forgot to change some, thanks

EFF commented on modules/qna/src/backend/storage.ts:
```diff
@@ -139,29 +150,103 @@ export default class Storage {
+    const { topicName, botId, zipFile, override, clean } = importArgs
+
+
+    if (!clean) {
```
This seems to be the complete inverse logic. if(clean) then remove all questions right @allardy

EFF commented:
Disable feature does not work for real
<img width="1176" alt="Screen Shot 2020-08-25 at 10 38 41 AM" src="https://user-images.githubusercontent.com/955524/91188520-62dcc900-e6bf-11ea-9fe5-2c8cee39107e.png">

Commit:
test filepicker

Ierezell commented on modules/qna/src/views/full/Components/ImportModal.tsx:
```diff
@@ -236,14 +153,14 @@ export const ImportModal: FC<Props> = props => {
   return (
     <Fragment>
       <Dialog
```
Hacked the button with a false click on an <input> with useRef

Ierezell commented on modules/qna/src/backend/storage.ts:
```diff
@@ -139,29 +150,103 @@ export default class Storage {
+      file: path.join(tmpDir.name, `${topicName}.tar.gz`)
+    })
+
+    const qnas = JSON.parse(await fse.readFile(path.join(tmpDir.name, FLOW_FOLDER, 'exportable.qna.intents.json'), 'utf8'))
```
The error is catched one level higher, I added extensive validation

Commit:
made revisions

Commit:
added filter & input file

Ierezell commented:
Last changes made : 

By default only one prompt : "This will override previous qna are you sure" <= No clean insert & Override only

Filtered Qna which are disabled => Creating intent only with enabled qna 

Changed the import button so it's a filepicker @MaxCloutier can you add a grey :hover and check if it's correct ?

Commit:
fixed disabling bug

Commit:
fixed qna replying to disable

MaxCloutier committed:
Fixed styling for QnA import

MaxCloutier commented:
@Ierezell just added the styling for hover and moved the styling to the stylesheet

MaxCloutier commented on modules/qna/src/views/full/index.tsx:
```diff
@@ -1,15 +1,15 @@
-import { ImportModal } from './Components/ImportModal'
 import QnA from './Components/QnA'
 import EmptyStateIcon from './Icons/EmptyStateIcon'
+import { props } from 'bluebird'
```
Umm is that normal? 🤔

Ierezell replied:
I removed it because it wasn't used

Commit:
problem came from old dev version

allardy committed:
Merge branch 'dev' into sp_refac-qna

Commit:
Merge branch 'sp_refac-qna' of github.com:botpress/botpress into ps_fix_sp_refac_qna

Commit:
rechecker prettier

Ierezell commented:
with the pull from sp_refac-qna which was pulled  with dev, everything works fine, qna can be disabled and thus will not be responded anymore

Commit:
added filtering

Commit:
added math magic for sortBy

Commit:
Merge branch 'dev' of github.com:botpress/botpress into ps_fix_sp_refac_qna

Commit:
fixed highlighting

Ierezell commented:
Resuming all the work : 

* Backend : 
    - Added routes for : 
        - fetching all questions 
        - fetch one question 
        - modify one question 
        - delete one question 
        - get actions corresponding to the qna (and modify old calls)
        - Export / Import from .tar.gz so you keep the content (images etc...)
        - Get number of questions  for tooltips 
    - Storage : 
        - Corresponding to the routes
* Frontend : 
    - Changed the import button => direct filepicker with only one confirmation ("Are you sure, it will override")
    - Added disabling question 
    - Added filter by active, disabled and incomplete
    - Added sortBy (most or least) recent

Commit:
reruned prettier

allardy commented on modules/qna/src/backend/storage.ts:
```diff
@@ -33,271 +33,257 @@ const normalizeQuestions = (questions: string[]) =>
+    .split('/')
+    .slice(-2)
+    .join('/')
+const removeBotPrefix = c =>
```
Missing typings

allardy commented on modules/qna/src/backend/storage.ts:
```diff
@@ -33,271 +33,257 @@ const normalizeQuestions = (questions: string[]) =>
-    const deletePromise = async (id: string): Promise<void> => {
-      await this.deleteMatchingIntent(id)
-      return this.bp.ghost.forBot(this.botId).deleteFile(this.config.qnaDir, `${id}.json`)
+    for (const mediaFile of fse.readdirSync(path.join(tmpDir.name, MEDIA_FOLDER))) {
```
`await fse.readdir`

Also, you don't check if the folder exists prior to reading it. If you don't have any media, then the folder will not be created in the archive and it will fail

allardy commented on modules/qna/src/translations/en.json:
```diff
@@ -88,6 +90,7 @@
     "uploadFile": "Upload File",
     "uploadStatus": "Upload status",
     "uploadSuccessful": "Upload successful",
+    "uploading": "Uploading",
```
Can you put single words translations in `ui-shared` instead ? We will just end up translating them multiple times otherwise.

allardy commented on modules/qna/src/views/full/Components/QnA.tsx:
```diff
@@ -219,20 +219,20 @@ const QnA: FC<Props> = props => {
               bp={bp}
             />
-          )}
+          )} */}
```
Commented, should it be removed?

Ierezell replied:
I guess so, it was old word from maxime and we don't use context anymore....

allardy commented on modules/qna/src/views/full/index.tsx:
```diff
@@ -23,12 +23,12 @@ const QnAList: FC<Props> = ({
   const [currentLang, setCurrentLang] = useState(contentLang)
   const [url, setUrl] = useState('')
+  const [filterOptions, setFilterOptions] = useState({ disabled: true, incomplete: true, active: true })
+  const [sortOption, setSortOption] = useState(+1)
```
Can you put this in the reducer instead?  Also, the sort option feels weird. Can you use a string value instead ? +1 / -1 doesn't mean anything

allardy commented on modules/qna/src/views/full/index.tsx:
```diff
@@ -156,18 +133,53 @@ const QnAList: FC<Props> = ({
+          label: lang.tr('leastRecent'),
+          selected: sortOption === -1,
+          action: () => {
+            setSortOption(-1)
```
sortOption = 'lastModifiedAsc' | 'lastModifiedDesc' 

allardy commented on modules/qna/src/views/full/index.tsx:
```diff
@@ -156,18 +133,53 @@ const QnAList: FC<Props> = ({
+          }
+        },
+        {
+          label: lang.tr('leastRecent'),
```
Translations are missing the `module.qna` prefix

allardy commented on modules/qna/src/views/full/index.tsx:
```diff
@@ -176,56 +188,94 @@ const QnAList: FC<Props> = ({
+      icon: 'export',
+      disabled: !items.length,
+      onClick: startDownload,
+      tooltip: noItemsTooltip || lang.tr('exportToJson')
```
Technically it's exported to an archive, not JSON anymore

allardy commented on modules/qna/src/views/full/index.tsx:
```diff
@@ -176,56 +188,94 @@ const QnAList: FC<Props> = ({
+    }
+  }
+
+  const importTar = async file => {
```
importArchive. Maybe we could also import JSON for backward compatibility?

allardy commented on modules/qna/src/views/full/index.tsx:
```diff
@@ -176,56 +188,94 @@ const QnAList: FC<Props> = ({
+        }
+      }, 500)
+    } catch (err) {
+      toastFailure(err.message)
```
The interval is not cleared if there's an error

allardy commented on src/bp/core/services/nlu/intent-service.ts:
```diff
@@ -42,15 +49,15 @@ export class IntentService {
             filename: flow.name,
             name: intentName,
-            slots: flow.variables?.map(x => ({ name: x.params?.name, entity: x?.params?.subType ?? x.type })) ?? [],
+            slots: flow.variables?.map(x => ({ name: x.params?.name, entity: x.type })) ?? [], // TODO: sub-type (type of list entity and type of pattern is not included in variables yet)
```
Here `x.type` will be "enum" and you will not have the correct name of the entity. It really is subType

allardy commented on modules/qna/src/views/full/index.tsx:
```diff
@@ -176,56 +188,94 @@ const QnAList: FC<Props> = ({
+            }}
+          />
+        </span>
+      ),
```
You don't need to put the input field next to the button. Instead, move the field somewhere in the page, then use a reference to it to trigger the click event

```js
const uploadRef = useRef()

{
  onClick: () => uploadRef?.current?.click()
}

// Put this somewhere in the file
<input
  type="file"
  ref={uploadRef}
            onChange={async e => {
              if ((e.target as HTMLInputElement).files) {
                await askUploadOptions((e.target as HTMLInputElement).files[0])
              }
            }}
          />
```

allardy commented on src/bp/ui-shared/src/MainContent/Header/typings.d.ts:
```diff
@@ -11,6 +11,7 @@ export interface HeaderProps {
 export interface HeaderButtonProps {
   onClick?: () => void
   icon?: IconName
+  content?: any
```
Please remove this, there is a better alternative

allardy commented on src/bp/ui-studio/src/web/views/OneFlow/sidePanel/Library/Modal.tsx:
```diff
@@ -0,0 +1,30 @@
+    </Dialog.Wrapper>
+  )
+}
+
```
This should be deleted, old stuff

allardy commented on src/bp/ui-studio/src/web/views/OneFlow/sidePanel/Library/NameModal.tsx:
```diff
@@ -0,0 +1,42 @@
+  )
+}
+
+export default EntityNameModal
```
Also old stuff, not used anymore

Ierezell commented on modules/qna/src/backend/api.ts:
```diff
@@ -1,186 +1,167 @@
+        botId: req.params.botId,
+        zipFile: req.file.buffer
+      }
+      await storage.importPerTopic(importArgs)
```
Migration will be hanlded in migration tools but the code is done

Commit:
corrected comments yann

Commit:
Merge branch 'dev' of github.com:botpress/botpress into ps_fix_sp_refac_qna

Commit:
fixed comments

Commit:
added typings to util function

Commit:
added autorefresh

Commit:
runed prettier

Commit:
Merge branch 'dev' of github.com:botpress/botpress into ps_fix_sp_refac_qna

Commit:
Merge branch 'dev' of github.com:botpress/botpress into ps_fix_sp_refac_qna

Commit:
fixed merge conflicts

EFF commented:
Filter and sort should be disabled when no item

EFF commented:
Import should not be visible when being readonly (not in default language)

EFF commented:
![Screen Shot 2020-09-03 at 12 03 53 PM](https://user-images.githubusercontent.com/955524/92139296-8bf70b00-eddd-11ea-83cc-9119a1599ccb.png)

EFF commented:
<img width="289" alt="Screen Shot 2020-09-03 at 12 07 50 PM" src="https://user-images.githubusercontent.com/955524/92141127-06c12580-ede0-11ea-9421-16905ed0c693.png">

EFF commented:
when highlighting a qna

![Screen Shot 2020-09-03 at 12 32 00 PM](https://user-images.githubusercontent.com/955524/92142220-7edc1b00-ede1-11ea-91af-cbf4400d401f.png)

Commit:
Merge branch 'dev' of github.com:botpress/botpress into ps_fix_sp_refac_qna

Commit:
changes and merge done

EFF committed:
Merge branch 'ps_fix_sp_refac_qna' of github.com:botpress/botpress into pierre_qna

Commit:
test with eff

EFF committed:
Merge branch 'ps_fix_sp_refac_qna' of github.com:botpress/botpress into pierre_qna

Commit:
delete functional

EFF committed:
Merge branch 'ps_fix_sp_refac_qna' of github.com:botpress/botpress into pierre_qna

EFF committed:
fixed update newly created items

EFF committed:
Merge branch 'dev' of github.com:botpress/botpress into pierre_qna
